### ISSUE TITLE
Invalid syntax is not well sanitized, compilation crashes with very general `Debug Failure`

### ISSUE THREAD

**Issue by retep007** [Microsoft/TypeScript#17548]
**TypeScript Version:**  2.4.2

**Code**

```ts
let headings: [][string] = [];
```

**Expected behavior:**
Shouldn't pass through syntax checking
**Actual behavior:**
Fails miserably with debug exception
```
/usr/local/lib/node_modules/typescript/lib/tsc.js:55776
                throw e;
                ^

Error: Debug Failure.
    at typeToString (/usr/local/lib/node_modules/typescript/lib/tsc.js:23526:22)
    at getPropertyTypeForIndexType (/usr/local/lib/node_modules/typescript/lib/tsc.js:27875:104)
    at getIndexedAccessType (/usr/local/lib/node_modules/typescript/lib/tsc.js:27924:20)
    at getTypeFromIndexedAccessTypeNode (/usr/local/lib/node_modules/typescript/lib/tsc.js:27929:38)
    at checkIndexedAccessType (/usr/local/lib/node_modules/typescript/lib/tsc.js:36181:41)
    at checkSourceElement (/usr/local/lib/node_modules/typescript/lib/tsc.js:38723:28)
    at checkVariableLikeDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:37122:13)
    at checkVariableDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:37213:20)
    at checkSourceElement (/usr/local/lib/node_modules/typescript/lib/tsc.js:38763:28)
    at Object.forEach (/usr/local/lib/node_modules/typescript/lib/tsc.js:298:30)
```

**Comment by ikatyang** [comment]
Should fixed by #17311.

(typescript@next)

```ts
let headings: [][string] = [];
//               ^^^^^^ [ts] Type '[]' has no matching index signature for type 'string'.
```

**Comment by retep007** [comment]
@ikatyang thank you for quick response, closing issue
